documentation/adr/002-trust-github-oidc.md

```diff
@@ -0,0 +1,58 @@
+---
+layout: page.njk
+status: Accepted
+date: git Created
+tags:
+  - adr
+order: 2
+title: 002 - Trust GitHub through OpenID Connect for platform development 
+---
+
+**Date: {{ page.date.toISOString().substring(0,10) }}**
+
+**Status: {{ status }}**
+
+## Context
+
+We have decided in [ADR 1](001-use-aws-landing-zone-accelerator.md) to use LZA.
+
+Each LZA deployment needs source code in a CodeCommit repository in the same account.
+
+We have multiple root accounts each containing an LZA deployment, and we want to be able to manage those environments from a single source code repository.
+
+We need to identify a platform for our developers to maintain our infrastructure and tooling code.
+
+Using CodeCommit is not desirable as we would have to share repositories across different root accounts.
+
+GitHub Enterprise is the Home Office's strategic developer platform.
+
+Other platforms such as GitLab, Code Commit and Bitbucket are also used in the Home Office.
+
+One of our principles is 'Strategic First', which means we should defer to the Home Office strategic solution by default.
+
+This ADR is only concerned with our platform team and is not a decision or limitation for customer teams.
+
+OpenID Connect (OIDC) is an open authentication protocol that is supported by AWS and GitHub.
+
+There needs to be a process where trust can be revoked at short notice in case of any emergent security concerns.
+
+## Decision
+
+We will use GitHub to maintain our infrastructure and tooling code.
+
+We will use OpenId Connect to establish a trusted relationship.
+
+We will set up one OIDC in each LZA management account and scope IAM roles to our code repository.
+
+We will update the CodeCommit repositories from GitHub in a write-only way.
```

Comment:
I'm not sure this is true, we will have to have read-only permissions. But it will be a one way write-only way with CodeCommit being updated by GitHub and not the other way around.

Reply:
@LiamMacP if it's a push from GitHub to CodeCommit then I don't think CodeCommit will need read access to the repo?

Whereas a CodeCommit doing a pull would absolutely need read access.

Reply:
I may be misunderstanding it and getting it the wrong way around. But the GHA from GitHub will need to pull (or more likely query) the CodeCommit repository to be able to push I think...

It may be that I'm misunderstanding and getting the relationship this is outlining the wrong way around though...

Reply:
Ah apologies, I thought you meant CodeCommit would have read access to the GH Repo

Reply:
I should have been more descriptive, my suggestion is something like: 

'GitHub will need pull permissions to be able to query the repository to coordinate the appropriate push actions.'

Reply:
I think is about the audience, that's too implementation specific for me. Do we have to state that there will be a TCP connection and some packets will be returned with data making it technically not write only? or is what's important from the platform operators perspective the effect of the decision is that CodeCommit is write only.

Reply:
Decision: There will need to be a process to manage data flow from Github to CodeCommit
@stephenkaye-ho @LiamMacP

Reply:
that will follow principle of least privilege